BaragonService/src/main/java/com/hubspot/baragon/service/managers/ServiceManager.java

```diff
@@ -25,6 +26,25 @@ public ServiceManager(BaragonStateDatastore stateDatastore, RequestManager reque
     this.requestManager = requestManager;
   }
 
+  public Optional<BaragonServiceState> getService(String host, String port) {
+    Collection<String> services = stateDatastore.getServices();
+
+    try {
+      for (String serviceId : services) {
+        Collection<UpstreamInfo> upstreams = stateDatastore.getUpstreams(serviceId);
+        for (UpstreamInfo upstreamInfo : upstreams) {
+          String upstream = upstreamInfo.getUpstream();
+          if (upstream.matches(String.format("%s\\.\\w+:%s", host, port))) {
+            return getService(serviceId);
+          }
+        }
+      }
+    }catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+    return Optional.absent();
+  }
```

Comment:
This seems horribly inefficient, but can't think of anything better since zk is backing it. Unless there's another place this data is stored I don't know about?